Panic in handleChainUpdates

Running dcrpool eb0117540746c7dc116b380590cc3781471d90be with current dcrd and dcrwallet master

```
panic: runtime error: index out of range [2] with length 2

goroutine 97 [running]:
github.com/decred/dcrpool/pool.(*ChainState).handleChainUpdates(0xc0002f0a50, 0xe5b0e0, 0xc0002c8900)
        /home/jholdstock/code/jholdstock/dcrpool/pool/chainstate.go:423 +0x2145
created by github.com/decred/dcrpool/pool.(*Hub).Run
        /home/jholdstock/code/jholdstock/dcrpool/pool/hub.go:565 +0xec
```

Here is the offending line:

https://github.com/decred/dcrpool/blob/eb0117540746c7dc116b380590cc3781471d90be/pool/chainstate.go#L423

`spew.Dump(block.Transactions[0])` gives the following:

```
(*wire.MsgTx)(0xc0000a6320)({
 CachedHash: (*chainhash.Hash)(<nil>),
 SerType: (wire.TxSerializeType) 0,
 Version: (uint16) 3,
 TxIn: ([]*wire.TxIn) (len=1 cap=1) {
  (*wire.TxIn)(0xc0005404e0)({
   PreviousOutPoint: (wire.OutPoint) 0000000000000000000000000000000000000000000000000000000000000000:4294967295,
   Sequence: (uint32) 4294967295,
   ValueIn: (int64) 30000000000,
   BlockHeight: (uint32) 0,
   BlockIndex: (uint32) 4294967295,
   SignatureScript: ([]uint8) (len=8 cap=8) {
    00000000  00 00 2f 64 63 72 64 2f                           |../dcrd/|
   }
  })
 },
 TxOut: ([]*wire.TxOut) (len=2 cap=2) {
  (*wire.TxOut)(0xc0000a6410)({
   Value: (int64) 0,
   Version: (uint16) 0,
   PkScript: ([]uint8) (len=14 cap=14) {
    00000000  6a 0c 23 00 00 00 07 af  74 5d 3d 5d dc 49        |j.#.....t]=].I|
   }
  }),
  (*wire.TxOut)(0xc0000a6438)({
   Value: (int64) 30000005770,
   Version: (uint16) 0,
   PkScript: ([]uint8) (len=25 cap=25) {
    00000000  76 a9 14 23 d4 15 0e b4  33 27 33 b5 bf 88 e5 d9  |v..#....3'3.....|
    00000010  ce a3 89 7b c0 9d bc 88  ac                       |...{.....|
   }
  })
 },
 LockTime: (uint32) 0,
 Expiry: (uint32) 0
})
```

--- a/go.mod
+++ b/go.mod
@@ -13,7 +13,7 @@ require (
 	github.com/decred/dcrd/dcrutil/v3 v3.0.0-20200608124004-b2f67c2dc475
 	github.com/decred/dcrd/rpc/jsonrpc/types/v2 v2.0.1-0.20200503044000-76f6906e50e5
 	github.com/decred/dcrd/rpcclient/v6 v6.0.0-20200623174822-e2d77e4e7efe
-	github.com/decred/dcrd/wire v1.3.0
+	github.com/decred/dcrd/wire v1.4.0
 	github.com/decred/slog v1.0.0
 	github.com/gorilla/csrf v1.7.0
 	github.com/gorilla/mux v1.7.4

--- a/pool/chainstate.go
+++ b/pool/chainstate.go
@@ -9,13 +9,14 @@ import (
 	"sync"
 	"sync/atomic"
 
+	"github.com/decred/dcrd/blockchain/standalone"
 	"github.com/decred/dcrd/chaincfg/chainhash"
 	"github.com/decred/dcrd/dcrutil/v3"
 	"github.com/decred/dcrd/wire"
 	bolt "go.etcd.io/bbolt"
 )
 
-var (
+const (
 	// bufferSize represents the block notification buffer size.
 	bufferSize = 128
 )
@@ -28,7 +29,7 @@ type ChainStateConfig struct {
 	// SoloPool represents the solo pool mining mode.
 	SoloPool bool
 	// PayDividends pays mature mining rewards to participating accounts.
-	PayDividends func(context.Context, uint32) error
+	PayDividends func(context.Context, uint32, bool) error
 	// GeneratePayments creates payments for participating accounts in pool
 	// mining mode based on the configured payment scheme.
 	GeneratePayments func(uint32, *PaymentSource, dcrutil.Amount, int64) error
@@ -269,6 +270,18 @@ func (cs *ChainState) pruneJobs(height uint32) error {
 	return err
 }
 
+// isTreasuryActive checks the provided coinbase transaction if
+// the treasury agenda is active.
+func isTreasuryActive(tx *wire.MsgTx) bool {
+	if !standalone.IsCoinBaseTx(tx) {
+		return false
+	}
+	if tx.Version < wire.TxVersionTreasury {
+		return false
+	}
+	return true
+}
+
 // handleChainUpdates processes connected and disconnected block
 // notifications from the consensus daemon.
 func (cs *ChainState) handleChainUpdates(ctx context.Context) {
@@ -333,8 +346,24 @@ func (cs *ChainState) handleChainUpdates(ctx context.Context) {
 				}
 			}
 
+			block, err := cs.cfg.GetBlock(ctx, &header.PrevBlock)
+			if err != nil {
+				// Errors generated fetching blocks of confirmed mined
+				// work are curently fatal because payments are
+				// sourced from coinbases. The chainstate process will be
+				// terminated as a result.
+				log.Errorf("unable to fetch block with hash %x: %v",
+					header.PrevBlock, err)
+				close(msg.Done)
+				cs.cfg.Cancel()
+				continue
+			}
+
+			coinbaseTx := block.Transactions[0]
+			treasuryActive := isTreasuryActive(coinbaseTx)
+
 			// Process mature payments.
-			err = cs.cfg.PayDividends(ctx, header.Height)
+			err = cs.cfg.PayDividends(ctx, header.Height, treasuryActive)
 			if err != nil {
 				log.Errorf("unable to process payments: %v", err)
 				close(msg.Done)
@@ -399,28 +428,21 @@ func (cs *ChainState) handleChainUpdates(ctx context.Context) {
 					continue
 				}
 
-				// TODO: look into keeping track of payment processing for
-				// confirmed mined work to facilitate recovery when the
-				// getblock calls err.
-				block, err := cs.cfg.GetBlock(ctx, &header.PrevBlock)
-				if err != nil {
-					// Errors generated fetching blocks of confirmed mined
-					// work are curently fatal because payments are
-					// sourced from coinbases. The chainstate process will be
-					// terminated as a result.
-					log.Errorf("unable to fetch block with hash %x: %v",
-						header.PrevBlock, err)
-					close(msg.Done)
-					cs.cfg.Cancel()
-					continue
-				}
-
 				// Generate payments for the confirmed block.
 				source := &PaymentSource{
 					BlockHash: block.BlockHash().String(),
-					Coinbase:  block.Transactions[0].TxHash().String(),
+					Coinbase:  coinbaseTx.TxHash().String(),
 				}
-				amt := dcrutil.Amount(block.Transactions[0].TxOut[2].Value)
+
+				// The coinbase output prior to
+				// [DCP0006](https://github.com/decred/dcps/pull/17)
+				// activation is at the third index position and at
+				// the second index position once DCP0006 is activated.
+				amt := dcrutil.Amount(coinbaseTx.TxOut[1].Value)
+				if !treasuryActive {
+					amt = dcrutil.Amount(coinbaseTx.TxOut[2].Value)
+				}
+
 				err = cs.cfg.GeneratePayments(block.Header.Height, source,
 					amt, work.CreatedOn)
 				if err != nil {

--- a/pool/chainstate_test.go
+++ b/pool/chainstate_test.go
@@ -38,7 +38,7 @@ func testChainState(t *testing.T, db *bolt.DB) {
 		t.Fatalf("unexpected serialization error: %v", err)
 	}
 
-	payDividends := func(context.Context, uint32) error {
+	payDividends := func(context.Context, uint32, bool) error {
 		return nil
 	}
 	generatePayments := func(uint32, *PaymentSource, dcrutil.Amount, int64) error {

--- a/pool/paymentmgr.go
+++ b/pool/paymentmgr.go
@@ -725,7 +725,7 @@ func (pm *PaymentMgr) maturePendingPayments(height uint32) (map[string][]*Paymen
 //
 // TODO: need to break this down into smaller funcs to make it easier to
 // test.
-func (pm *PaymentMgr) payDividends(ctx context.Context, height uint32) error {
+func (pm *PaymentMgr) payDividends(ctx context.Context, height uint32, treasuryActive bool) error {
 	pmts, err := pm.maturePendingPayments(height)
 	if err != nil {
 		return err
@@ -779,7 +779,15 @@ func (pm *PaymentMgr) payDividends(ctx context.Context, height uint32) error {
 	}
 
 	for _, set := range pmts {
-		index := uint32(2)
+		// The coinbase output prior to
+		// [DCP0006](https://github.com/decred/dcps/pull/17)
+		// activation is at the third index position and at
+		// the second index position once DCP0006 is activated.
+		index := uint32(1)
+		if !treasuryActive {
+			index = 2
+		}
+
 		txHash, err := chainhash.NewHashFromStr(set[0].Source.Coinbase)
 		if err != nil {
 			return err
@@ -788,6 +796,11 @@ func (pm *PaymentMgr) payDividends(ctx context.Context, height uint32) error {
 		txOutResult, err := txCreator.GetTxOut(ctx, txHash, index, false)
 		if err != nil {
 			return fmt.Errorf("unable to find tx output: %v", err)
+		}
+
+		if txOutResult == nil {
+			return fmt.Errorf("no transaction output found for hash %s "+
+				"at index %d", txHash, index)
 		}
 
 		// Ensure the referenced prevout to be spent is a coinbase and

